<launch>

<!--Intel Realsense Camera Image + IMU launch code-->
    <!--In order to launch another .launch file from different package, use <include> tag-->
    <include file="$(find realsense2_camera)/launch/rs_camera.launch">
        <arg name="unite_imu_method" value='linear_interpolation'/>
        <arg name="publish_tf" value="false"/>
        <arg name="publish_odom_tf" value="false"/>
	<arg name="enable_infra1" value="false"/>
	<arg name="enable_infra2" value="false"/>
	<arg name="enable_depth" value="false"/>
    </include>

<!--RPLidar node launch code-->
    <node name="rplidarNode" pkg="rplidar_ros" type="rplidarNode" output="screen">
        <!-- Change the device setting parameters according to current HW state -->
        <param name="serial_port" type="string" value="/dev/ttyUSB0"/>
        <param name="serial_baudrate" type="int" value="115200"/><!--A1/A2 -->
        <!--param name="serial_baudrate"     type="int"    value="256000"--><!--A3 -->

        <!-- *** Set frame_id as base_laser in order to bind RPLidar to base_laser TF ***-->
        <param name="frame_id" type="string" value="base_laser"/>   
        <!-- *** ROS and gmapping requires the location (TF) of Lidar in order to create the map *** -->
        <!-- *** As a result, it is important to bind RPLidar node to base_laser TF by assigning frame_id as base_laser *** -->
        
        <param name="inverted" type="bool" value="false"/>
        <param name="angle_compensate" type="bool" value="true"/>
    </node>
    <node name="lidar_node" pkg="slam_proto_v1" type="lidar_node">
    </node>

<!--USB Cam node launch code-->
<!--
    <node name="usb_cam_node" pkg="usb_cam" type="usb_cam_node">
-->
        <!-- Change the device setting parameters according to current HW state -->
        <!-- In order to receive video stream from camera, appropriate format and codec settings have to be applied -->
<!--
        <param name="video_device" type="string" value="/dev/video2"/>
        <param name="pixel_format" type="string" value="yuyv"/>
    </node>
    <node name="cam_node" pkg="slam_proto_v1" type="cam_node">
    </node>
-->

<!--Madgwick IMU Filter launch code-->
    <node name="imu_filter_madgwick" pkg="imu_filter_madgwick" type="imu_filter_node">
        <param name="use_mag" type="bool" value="false"/>
        <param name="_publish_tf" type="bool" value="false"/>
        <param name="_world_frame" type="string" value="enu"/>
        <remap from="imu/data_raw" to="camera/imu"/>
    </node>    

<!--Serial Communication node launch code-->
    <!--To launch python-based node, 'type' attribtue in <node> has to be 'filename.py'. This is due to ROS naming convetion.-->
    <node name="serial_port" pkg="rosserial_python" type="serial_node.py" output="screen">
        <!-- Change the device setting parameters according to current HW state -->
        <param name="port" value="/dev/ttyACM0"/>
        <param name="baud" value="57600"/>
        <!-- To adjust connection settings for rosserial, we have to access the setting values as 'param' attributes. -->
    </node>
    <node name="serial_RX_node" pkg="slam_proto_v1" type="serial_RX_node">
    </node>

<!--Transformation node launch code-->
<!--
    <node name="tf_broadcaster" pkg="slam_proto_v1" type="tf_broadcaster">
    </node>
-->
<!--Gmapping node launch code-->
    <node name="slam_gmapping" pkg="gmapping" type="slam_gmapping">
        <param name="scan" value="/scan"/>
        <param name="map_update_interval" value="2"/>
        <param name="particles" value="20"/>
	<param name="delta" value="0.1"/>
        <param name="linearUpdate" value="0.1"/>
        <param name="angularUpdate" value="0.5"/>
    </node>

<!--Ackermann Odom node launch code-->

    <node name="ackermann_odom_broadcaster" pkg="slam_proto_v1" type="ackermann_odom_broadcaster" output="screen">
    </node>
    
</launch>
